### ISSUE TITLE
Improve docs on async methods?

### ISSUE THREAD

**Issue by peterharperuk** [raspberrypi/pico-feedback#311]
Wifi support in Pico W could use different methods depending on the users choice, polling or background threadsafe. We could probably do with something to describe the thinking behind this. The latest sdk has a new generic asynchronous API which behaves the same.

Basically if you're a user and you need to use wifi or bluetooth or a network stack or need to do something with the async api you need to make a decision about which method you use and what the implications are. Freertos complicates matters even further.

I'm wondering if we need a separate section in the docs to go into this? It's hard to gain an understanding from the  API documentation.

**Comment by lurch** [comment]
Sounds like content for "Connecting to the Internet with Raspberry Pi Pico W" ? So I've moved it to the pico-feedback repo.

**Comment by peterharperuk** [comment]
The same issue affects bluetooth or the new async_context API which you could use for anything. It's more of a "driver" thing. An interrupt goes off (e.g. to tell us the wifi chip needs attention) and we have to do the work "somewhere else". This has implications / trade-offs for the user. Maybe if I write something you can knock it into shape and decide where it lives?

**Comment by lurch** [comment]
Okay, so perhaps it belongs as a new chapter in https://datasheets.raspberrypi.com/pico/raspberry-pi-pico-c-sdk.pdf then?
Ultimately it's Alasdair's decision to make, so I'll assign this issue to him (for now). But yeah, I'm sure any notes / documentation you can come up with will be greatly appreciated :+1:  Thanks!

**Comment by ghazanhaider** [comment]
Adding my voice to this.
The runtime scheduler and async context are not well described for users with previous embedded background and users who want lower level control of even flows. The hardware_timer and datasheets also do not describe this aspect of the runtime.

**Comment by nathan-contino** [comment]
Is this still an issue? If so, I can prioritise it sometime later this year.

**Comment by peterharperuk** [comment]
Yes, probably. It was really a reminder for me to write something, which of course never happened.

**Comment by mjcross** [comment]
@peterharperuk @nathan-contino  I can't say enough how valuable this would be.... _pleeeese!_ As a poor dumb user, trying to understand how to start using the asynch methods feels a bit like being blindfolded and thrown down a coal mine. Trying to work it out from the API documentation for lwip and/or cy43_arch without a high level introduction to know which methods to use is just _painful_. I think the mistake may be trying to make such documentation too comprehensive. May I suggest doing it a bit piece by piece - perhaps with one of the existing Pico_W examples torn down and explained with a bit of hand-holding? Otherwise I can see this still being ongoing for another two years :-)
